clang/test/SemaSYCL/sycl-esimd-inline-asm-err.cpp

```diff
@@ -0,0 +1,17 @@
+// RUN: %clang_cc1 -fsycl-is-device -fsyntax-only -verify %s
+
+// This check verifies the data_ref function on simd_obj_impl errors if not used in inline asm
+struct [[__sycl_detail__::sycl_type(esimd_simd_or_simd_mask)]] simd_obj_impl {
+   int& data_ref() { return x; }
+private:
+int x = 5;
+};
+
+
+__attribute__((sycl_device)) void func() {
+ simd_obj_impl obj;
+
+// expected-error@+1 {{Function may only be used in inline assembly}}
+int& loc = obj.data_ref();
+__asm__("%0" : "=r"(obj.data_ref()));
```

Comment:
Yes, there is not a correctness requirement, it will work fine with your example. However our team decided we don't want to give users a reference to the underlying vector, and we are only doing it because it's required for doing assignment in inline asm. So it's a design restriction not a technical one

Reply:
I spoke to the CFE architects about this diagnostic and there are many concerns. The design is problematic as it does not fully consider all the ways this function can be called. For example, what about calls through function pointers.  Also what stops users from using inline ASM in their own function to smuggle the reference out? A diagnostic which correctly/fully handles this is not really feasible in the frontend @erichkeane @premanandrao @AaronBallman

Reply:
IMO, it is still good to have such simple check.

It is like a small fence at the edge of a cliff on a hiking route fall-preventing those who walk and look at their phone, or who doesn't have idea where it stops being safe.
Surely, some still can jump over the fence, or steal the pointer returned by ref_data() via using asm, and  then hurt themselves, but that would be on their own risk/expense.
This error is a clear and visible message that the method is public, but is designed for limited usage - in asm only.

BTW, it can be a warning instead of an error as well.

Reply:
This sounds like a job for a static-analyzer for me.  Warnings on otherwise valid code that is guaranteed to do what they want, should not be warned on.

Reply:
@erichkeane - when you say "This sounds like a job for a static-analyzer for me". Do you mean this check needs to be somewhere in different place, not where it is now?

Reply:
Yes, a static analyzer is a separate tool from a compiler, like Klockworks or Coverity (or clang-tidy, or the Clang Static Analyzer, both of which are NOT the frontend).

Reply:
Static analyzers like Coverity will not find unintended uses of a public method ref_data() because it doesn't know which use is  good and which is not. Also, nobody runs Coverity-like tools for usual sycl apps.

The idea was to signal on too risky/unintended usage of the method before the code fails/hangs/etc at runtime.

If it is too expensive to do this check here, then Ok, I understand. Then, perhaps this check can be moved to some other place either in C++FE (deferred diagnostics?), or in the middle (sycl-post-link + use_def_analysis in one of passes like LowerESIMD).

Reply:
There is no place in the FE for this diagnostic and I'd be pretty strongly opposed to diagnosing it from the backend on usability grounds if the optimizer could end up changing the diagnostic behavior (such a diagnostic would not be upstreamed to Clang/LLVM, most likely). As Erich said, there are two places in the toolchain stack where this could potentially live: either as a static analyzer check or in a clang-tidy check. I agree that no third-party static analysis tool is interested in SYCL currently, so that basically leaves the clang static analyzer (and clang-tidy) as the only choice there.

Reply:
so i have a local implementation of an assert in the middle end (inside an ESIMD pass run in sycl-post-link) that just errors with 'unsupported use of function' or something like that. the implementation is simple, does not compromise compile time or runtime performance, and i do not expect the optimizer to cause false positives. are you all okay with doing it this way? if so i can clean it up and post it. thanks for the discussion.

Reply:
well my understanding of @v-klochkov's point is that the function does work fine, but it's very easy for users to shoot themselves in the foot, so we wanted to add an error/warning to only allow it in tested cases. personally i don't mind dropping the error/warning, i think giving the warning in the documentation for the function is probably enough

Reply:
I'm surprised anyone would argue that using inline assembly is *less likely* to cause a user to shoot themselves in the foot, but at the end of the day, such a design decision is very unusual for C++ and should be a red flag that the design has issues. If this is too dangerous to call directly, it should not be exposed to users. If it needs to be exposed to users, you should remove the sharp edges so they can use it. But changing the toolchain to work around a difficult-to-use design feels like the cart driving the horse.

Reply:
@v-klochkov do you have a case in mind where the user could hurt themselves calling the function directly? if so, can we do anything do fix that?

Reply:
Ok, let's drop the check for now.